File: backend/app/schemas/card.py
from datetime import datetime
from typing import List

from pydantic import BaseModel, Field


class CardBase(BaseModel):
    front_content: str = Field(..., min_length=1, max_length=5000)
    back_content: str = Field(..., min_length=1, max_length=5000)
    ease_factor: float | None = Field(default=None, ge=1.3, le=3.0)
    interval: int | None = Field(default=None, ge=0)
    repetitions: int | None = Field(default=None, ge=0)
    next_review: datetime | None = None
    order_index: int | None = Field(default=None, ge=0)


class CardCreate(CardBase):
    deck_id: int


class CardUpdate(BaseModel):
    front_content: str | None = Field(default=None, min_length=1, max_length=5000)
    back_content: str | None = Field(default=None, min_length=1, max_length=5000)
    ease_factor: float | None = Field(default=None, ge=1.3, le=3.0)
    interval: int | None = Field(default=None, ge=0)
    repetitions: int | None = Field(default=None, ge=0)
    next_review: datetime | None = None
    order_index: int | None = Field(default=None, ge=0)


class CardResponse(CardBase):
    id: int
    deck_id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class CardListResponse(BaseModel):
    items: list[CardResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class CardBulkCreateItem(BaseModel):
    front_content: str = Field(..., min_length=1, max_length=5000)
    back_content: str = Field(..., min_length=1, max_length=5000)
    order_index: int | None = Field(default=None, ge=0)


class CardBulkCreateRequest(BaseModel):
    deck_id: int
    items: List[CardBulkCreateItem]


class CardBulkUpdateItem(BaseModel):
    id: int
    front_content: str | None = Field(default=None, min_length=1, max_length=5000)
    back_content: str | None = Field(default=None, min_length=1, max_length=5000)
    ease_factor: float | None = Field(default=None, ge=1.3, le=3.0)
    interval: int | None = Field(default=None, ge=0)
    repetitions: int | None = Field(default=None, ge=0)
    next_review: datetime | None = None
    order_index: int | None = Field(default=None, ge=0)


class CardBulkUpdateRequest(BaseModel):
    deck_id: int
    items: List[CardBulkUpdateItem]

File: frontend/src/services/card.ts
import { api } from "./api";

export interface Card {
  id: number;
  front_content: string;
  back_content: string;
  deck_id: number;
  ease_factor: number;
  interval: number;
  repetitions: number;
  next_review: string;
  order_index: number;
  created_at: string;
  updated_at: string;
}

export interface CardCreate {
  deck_id: number;
  front_content: string;
  back_content: string;
  ease_factor?: number;
  interval?: number;
  repetitions?: number;
  next_review?: string;
  order_index?: number;
}

export interface CardUpdate {
  front_content?: string;
  back_content?: string;
  ease_factor?: number;
  interval?: number;
  repetitions?: number;
  next_review?: string;
  order_index?: number;
}

export interface CardBulkCreateRequest {
  deck_id: number;
  items: Array<{
    front_content: string;
    back_content: string;
    order_index?: number;
  }>;
}

export interface CardBulkUpdateRequest {
  deck_id: number;
  items: Array<{
    id: number;
    front_content?: string;
    back_content?: string;
    ease_factor?: number;
    interval?: number;
    repetitions?: number;
    next_review?: string;
    order_index?: number;
  }>;
}

export interface CardListResponse {
  items: Card[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface CardListParams {
  deck_id: number;
  page?: number;
  page_size?: number;
  search?: string;
}

export const cardService = {
  async list(params: CardListParams): Promise<CardListResponse> {
    const response = await api.get<CardListResponse>("/cards", { params });
    return response.data;
  },

  async create(data: CardCreate): Promise<Card> {
    const response = await api.post<Card>("/cards", data);
    return response.data;
  },

  async createBulk(payload: CardBulkCreateRequest): Promise<Card[]> {
    const response = await api.post<Card[]>("/cards/bulk", payload);
    return response.data;
  },

  async update(id: number, data: CardUpdate): Promise<Card> {
    const response = await api.patch<Card>(`/cards/${id}`, data);
    return response.data;
  },

  async updateBulk(payload: CardBulkUpdateRequest): Promise<Card[]> {
    const response = await api.patch<Card[]>("/cards/bulk", payload);
    return response.data;
  },

  async delete(id: number): Promise<void> {
    await api.delete(`/cards/${id}`);
  },
};

